diff -ur grub-2.02~beta3_o/util/grub.d/10_linux.in grub-2.02~beta3_n/util/grub.d/10_linux.in
--- grub-2.02~beta3_o/util/grub.d/10_linux.in	2016-07-18 17:44:44.857108348 +0100
+++ grub-2.02~beta3_n/util/grub.d/10_linux.in	2016-07-18 17:43:18.184891235 +0100
@@ -46,8 +46,8 @@
 
 # btrfs may reside on multiple devices. We cannot pass them as value of root= parameter
 # and mounting btrfs requires user space scanning, so force UUID in this case.
-if [ "x${GRUB_DEVICE_UUID}" = "x" ] || [ "x${GRUB_DISABLE_LINUX_UUID}" = "xtrue" ] \
-    || ! test -e "/dev/disk/by-uuid/${GRUB_DEVICE_UUID}" \
+if  [ "x${GRUB_DEVICE_UUID}" = "x" ] || [ "x${GRUB_DISABLE_LINUX_UUID}" = "xtrue" ] \
+    || ( ! test -e "/dev/disk/by-uuid/${GRUB_DEVICE_UUID}" && [ "x${DURING_INSTALL}" != "x1" ] ) \
     || ( test -e "${GRUB_DEVICE}" && uses_abstraction "${GRUB_DEVICE}" lvm ); then
   LINUX_ROOT_DEVICE=${GRUB_DEVICE}
 else
