diff -ur grub-2.02~beta3_o/util/grub.d/10_linux.in grub-2.02~beta3/util/grub.d/10_linux.in
--- grub-2.02~beta3_o/util/grub.d/10_linux.in	2017-05-15 17:34:34.000000000 +0100
+++ grub-2.02~beta3/util/grub.d/10_linux.in	2017-05-15 18:29:00.703840039 +0100
@@ -26,7 +26,11 @@
 export TEXTDOMAIN=@PACKAGE@
 export TEXTDOMAINDIR="@localedir@"
 
-CLASS="--class gnu-linux --class gnu --class os --unrestricted"
+if [ -e /boot/grub2/user.cfg ] && grep -q GRUB2_PASSWORD /boot/grub2/user.cfg; then
+  CLASS="--class gnu-linux --class gnu --class os"
+else
+  CLASS="--class gnu-linux --class gnu --class os --unrestricted"
+fi
 
 if [ "x${GRUB_DISTRIBUTOR}" = "x" ] ; then
   OS="$(eval $(grep PRETTY_NAME /etc/os-release) ; echo ${PRETTY_NAME})"
