--- util/grub-mkrescue.c	2018-04-17 23:49:24.443474449 -0700
+++ util/grub-mkrescue.c_new	2018-04-18 00:01:00.972896685 -0700
@@ -497,7 +497,7 @@
   
   iso9660_dir = grub_util_make_temporary_dir ();
   grub_util_info ("temporary iso9660 dir is `%s'", iso9660_dir);
-  boot_grub = grub_util_path_concat (3, iso9660_dir, "boot", "grub");
+  boot_grub = grub_util_path_concat (3, iso9660_dir, "boot", "grub2");
   grub_install_mkdir_p (boot_grub);
   romdir = grub_util_path_concat (2, boot_grub, "roms");
   grub_util_mkdir (romdir);
@@ -596,7 +596,7 @@
 				    "i386-pc-eltorito", 0);
 
       xorriso_push ("-b");
-      xorriso_push ("boot/grub/i386-pc/eltorito.img");
+      xorriso_push ("boot/grub2/i386-pc/eltorito.img");
       xorriso_push ("-no-emul-boot");
       xorriso_push ("-boot-load-size");
       xorriso_push ("4");
@@ -831,8 +831,8 @@
       grub_install_copy_file (grub_chrp, bootx, 1);
       grub_install_mkdir_p (ppc_chrp);
       grub_install_copy_file (bisrc, bitgt, 1);
-      xorriso_link ("/System/Library/CoreServices/grub.elf", "/boot/grub/powerpc-ieee1275/core.elf");
-      xorriso_link ("/boot/grub/powerpc.elf", "/boot/grub/powerpc-ieee1275/core.elf");
+      xorriso_link ("/System/Library/CoreServices/grub.elf", "/boot/grub2/powerpc-ieee1275/core.elf");
+      xorriso_link ("/boot/grub2/powerpc.elf", "/boot/grub2/powerpc-ieee1275/core.elf");
       /* FIXME: add PreP */
       if (system_area == SYS_AREA_COMMON)
 	{
@@ -881,25 +881,25 @@
       xorriso_push ("-B");
       xorriso_push (",");
       xorriso_push ("--grub2-sparc-core");
-      xorriso_push ("/boot/grub/sparc64-ieee1275/core.img");
+      xorriso_push ("/boot/grub2/sparc64-ieee1275/core.img");
     }
 
   make_image_fwdisk (GRUB_INSTALL_PLATFORM_MIPS_ARC, "mips-arc", "mips-arc/core.img");
 
   if (source_dirs[GRUB_INSTALL_PLATFORM_MIPS_ARC])
     {
-      xorriso_link ("/boot/grub/mips-arc/grub", "/boot/grub/mips-arc/core.img");
-      xorriso_link ("/boot/grub/mips-arc/sashARCS", "/boot/grub/mips-arc/core.img");
-      xorriso_link ("/boot/grub/mips-arc/sash", "/boot/grub/mips-arc/core.img");
+      xorriso_link ("/boot/grub2/mips-arc/grub", "/boot/grub2/mips-arc/core.img");
+      xorriso_link ("/boot/grub2/mips-arc/sashARCS", "/boot/grub2/mips-arc/core.img");
+      xorriso_link ("/boot/grub2/mips-arc/sash", "/boot/grub2/mips-arc/core.img");
     }
   if (source_dirs[GRUB_INSTALL_PLATFORM_MIPS_ARC] && system_area == SYS_AREA_ARCS)
     {
       xorriso_push ("-mips-boot");
-      xorriso_push ("/boot/grub/mips-arc/sashARCS");
+      xorriso_push ("/boot/grub2/mips-arc/sashARCS");
       xorriso_push ("-mips-boot");
-      xorriso_push ("/boot/grub/mips-arc/sash");
+      xorriso_push ("/boot/grub2/mips-arc/sash");
       xorriso_push ("-mips-boot");
-      xorriso_push ("/boot/grub/mips-arc/grub");
+      xorriso_push ("/boot/grub2/mips-arc/grub");
     }
 
   make_image_fwdisk (GRUB_INSTALL_PLATFORM_MIPSEL_ARC, "mipsel-arc", "arc.exe");
